### PR TITLE
Add 6.0 release

### LINKED ISSUE #552: Create MongoDB 6.0 Docker image
MongoDB 6.0 is GA: https://www.mongodb.com/blog/post/mongodb-6-0-now-available https://github.com/mongodb/mongo/releases/tag/r6.0.0

There should be a docker image for it.

### PR DESCRIPTION
**Pull request by yosifkit**
fixes #552

### CODE REVIEW

**Comment by tianon** [comment]
I pushed https://github.com/docker-library/mongo/compare/1886739107c864c9a1b18e8f2d176acaf6f0e4af..091bdab11774adf475955fc46b28b4dc69fe89a2 which removes `Client` from the Windows builds (although as I type this, realizing I forgot to remove `mongo --version` too :see_no_evil:) -- next will be looking at the failing test and what we'll need to do to update it to `mongosh`.

**Comment by tianon** [comment]
Just to record my digging on the latest Windows failure, I _think_ this is responsible for the build failing (although the wmi log output makes that hard to be sure about):

```
MSI (s) (A0:1C) [17:32:29:801]: Executing op: ActionStart(Name=InstallCompassScript,Description=Installing MongoDB Compass... (this may take a few minutes),)
MSI (s) (A0:1C) [17:32:29:879]: Executing op: CustomActionSchedule(Action=InstallCompassScript,ActionType=1089,Source=BinaryData,Target=WixQuietExec64,CustomActionData="C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Bypass -Command "& '' ; exit $($Error.Count)")
MSI (s) (A0:F0) [17:32:29:882]: Invoking remote custom action. DLL: C:\Windows\Installer\MSI9E27.tmp, Entrypoint: WixQuietExec64
WixQuietExec64:  The expression after '&' in a pipeline element produced an object that was not valid. It must result in a command 
WixQuietExec64:  name, a script block, or a CommandInfo object.
WixQuietExec64:  At line:1 char:3
WixQuietExec64:  + & '' ; exit $($Error.Count)
WixQuietExec64:  +   ~~
WixQuietExec64:      + CategoryInfo          : InvalidOperation: (:String) , RuntimeException
WixQuietExec64:      + FullyQualifiedErrorId : BadExpression
WixQuietExec64:   
WixQuietExec64:  Error 0x80070001: Command line returned an error.
WixQuietExec64:  Error 0x80070001: QuietExec64 Failed
WixQuietExec64:  Error 0x80070001: Failed in ExecCommon method
```

From what I can tell, https://github.com/mongodb/mongo/blob/r6.0.0/src/mongo/installer/msi/wxs/Installer_64.wxs#L199-L203 is somehow expanding `[#InstallCompassScript]` to the empty string. :sweat_smile:

**Comment by tianon** [comment]
I think https://github.com/mongodb/mongo/blob/r6.0.0/src/mongo/installer/msi/wxs/FeatureFragment.wxs#L87-L93 and https://github.com/mongodb/mongo/blob/r6.0.0/src/mongo/installer/msi/wxs/BinaryFragment.wxs#L5-L7 are somehow relevant, and not working as expected. :see_no_evil:

**Comment by yosifkit** [comment]
Oops lol :see_no_evil:, the current windows install failure is our fault; the log we are digging through is what it looks like on success.

https://github.com/docker-library/mongo/blob/9c233432872113295fd9b5525c28b93e0c0fcdc0/Dockerfile-windows.template#L42-L46

**Comment by yosifkit** [comment]
Fixed the check for `mongo.exe` and fixed nanoserver to not try running `mongo` when it doesn't exist :laughing:

**Comment by tianon** [comment]
I've fixed the test in https://github.com/docker-library/official-images/pull/12836 (so once that merges, we can restart the builds here and validate that we're still good on old versions and fixed on the new one :+1:)

**Comment by yosifkit** [comment]
More `mongo` assumptions causing a failure:
https://github.com/docker-library/mongo/blob/9c233432872113295fd9b5525c28b93e0c0fcdc0/docker-entrypoint.sh#L340

**Comment by tianon** [comment]
Gah, that one has deeper implications than the one in our test suite. :sweat_smile: :see_no_evil:

Up until now, any `.js` files in `/docker-entrypoint-initdb.d` have been implied to be run by `mongo ...` (and any `.sh` files invoking `"${mongo[@]}"` have been able to assume that's running `mongo ...`), and us just quietly changing that to `mongosh`, even if we only do it for 6.0+, is a little bit of a breaking change that I think we need to spend a little time considering more carefully. :innocent:

It's inevitable that we have to do _something_, because `mongo` is literally gone now, but I think we need to spend some time thinking about _what_ we do instead of just knee-jerking `s/mongo/mongosh/` as my own personal tendency would be.  The fact that even our absolute bare-minimum, total-basics test ran into several rough edges in the conversion that required some minor modification makes me think that perhaps _just_ swapping is probably going to have some pretty gnarly downstream effect.

We've never been particularly happy with how this script is implemented and the sorts of hacks it has to do to even function remotely in a way that makes sense, and I have to stop and wonder if we can avoid making an already annoying-to-maintain script even _worse_ in this transition. :grimacing: :see_no_evil:

(My gut says probably not, but I'm trying to stay optimistic -- this entrypoint script is totally out of control already with custom behavior we maintain above and beyond what MongoDB upstream maintains, supports, or even recommends, which is why we avoided creating it for so long in the first place!)

**Comment by Romain7495** [comment]
Hello, do we have an update on the `6.0.0-focal` that is failing ? @tianon

File: versions.json [hunk 1]
```diff
@@ -230,5 +230,66 @@
       }
     },
     "version": "5.0.9"
+  },
+  "6.0": {
+    "changes": "https://jira.mongodb.org/issues/?jql=project%20%3D%20SERVER%20AND%20fixVersion%20%3D%20%226.0.0%22%20ORDER%20BY%20status%20DESC%2C%20priority%20DESC",
```
Comment (akshaybabloo) [inline versions.json:235]: This issues doesn't seem to exists. Is this an internal one?

Comment (yosifkit) [reply versions.json:235]: That comes directly from https://downloads.mongodb.org/current.json. it is not something we actively use; it is just a link taken from the upstream json.

**Comment by sepatel** [comment]
> Gah, that one has deeper implications than the one in our test suite. sweat_smile see_no_evil
> 
> Up until now, any `.js` files in `/docker-entrypoint-initdb.d` have been implied to be run by `mongo ...` (and any `.sh` files invoking `"${mongo[@]}"` have been able to assume that's running `mongo ...`), and us just quietly changing that to `mongosh`, even if we only do it for 6.0+, is a little bit of a breaking change that I think we need to spend a little time considering more carefully. innocent

To add some color, there are quite a few breaking changes in mongo 6 vs mongo 5 actually and things are going to break. With mongosh specifically for example, there is no more .save() method on the collection so every script whether it is `.js` or `.sh` break because a LOT of people were doing `.save()` in the past. Also, the need to support .sh based scripts in general went away completely since you have a full fledge nodejs 16 to play with now which can do everything including running fork processes and shell commands.

Even if you created something custom to run scripts and stuff like before, the contents of all those scripts would break because it isn't backwards compatible. There is a module you can install (mongocompat I think it was) that is suppose to enable it to be backwards compatible as well but it in and of itself is a bit of a source of problem as well.

So either you can include the legacy shell in your container package which then causes docker container version to do things that other versions including what is officially supported by cloud mongo to not work out of the box correctly (breaking things in a worse way i'd imagine), or you embrace the fact that 6.x means it has breaking changes from 5.x (just like how 5.x had breaking changes from 4.x) and that in order to upgrade, you will have to make changes to things.

My preference is to forget backwards compatibility, axe the `.sh` script processing stuff as well, and embrace the power of mongosh as the new way of doing things cleanly.

**Comment by sohaha** [comment]
Have you made any progress?

**Comment by shahyar** [comment]
It's a major release version. No one can expect to blindly upgrade their image to 6.0 and have it work. With that said, from a cursory glance of the state of things right now, I'd hope it would be working (or close to it) if we `s/ mongo / mongosh /` the whole pipeline.

But yeah, this does look tough to maintain. There seems to be a lot of handholding of Docker users in this code, for what seems to be to avoid running into potentially cryptic config/setup issues with MongoDB? Gut it all and let people prep for 6.0 the right (hard 🥲) way.

**Comment by xianzhuo-sky** [comment]
> My preference is to forget backwards compatibility, axe the `.sh` script processing stuff as well, and embrace the power of mongosh as the new way of doing things cleanly.

we need to move forward

**Comment by tianon** [comment]
I've added 3f1ee375ae5f25d5145490fe6a4b38b3dd814b4f here -- this is explicitly a breaking change as it removes _all_ the (very, very hacky) environment variable support in favor of users doing this "correctly".  I'm not thrilled about this because doing so "correctly" is quite complicated, but there are so many edge cases here that I really believe we did more harm than good by introducing them. :see_no_evil: :disappointed:

**Comment by tianon** [comment]
https://www.mongodb.com/compatibility/docker references `MONGO_INITDB_ROOT_USERNAME`/`MONGO_INITDB_ROOT_PASSWORD`, but https://www.mongodb.com/docs/manual/tutorial/install-mongodb-enterprise-with-docker/ and https://www.mongodb.com/compatibility/deploying-a-mongodb-cluster-with-docker both seem to explicitly _not_ reference any of our `MONGO_` variables (except for the `ARG` values we still include in the `Dockerfile`), so I think this is probably still in line with MongoDB's expectations for this image? :grimacing:

(I really wish we could find a good contact to ask questions like this to make sure we're representing them well :see_no_evil:)

**Comment by Wazbat** [comment]
> it removes _all_ the (very, very hacky) environment variable support in favor of users doing this "correctly"

Ah, so this would remove support for ENV variables like `MONGO_INITDB_ROOT_USERNAME`? I understand why this could be a difficult choice, and I completely understand why you might be worried about how this "represents" mongodb as a whole.

Setting things up with only a few environment variables is _easy_ after all, and it feels like one of mongo's selling points to new devs is that it's an easy to use database...

Is there noone on the mongodb who can weigh in on this?

**Comment by razvanavram** [comment]
For anyone trying to get mongo:6.0.0 **until a version is officially released** use this Dockerfile (along with entrypoint) [link](https://github.com/docker-library/mongo/pull/554/commits/0d6cd3dadbace5516116909022cab4be6ee4cc21)

Also, in Dockerfile edit with this:
```
RUN chmod 755 /usr/local/bin/docker-entrypoint.sh
ENTRYPOINT ["/usr/local/bin/docker-entrypoint.sh"]
```

**Comment by ehiggs** [comment]
@tianon is there a tracking issue for the tests?

**Comment by sepatel** [comment]
> > it removes _all_ the (very, very hacky) environment variable support in favor of users doing this "correctly"
> 
> Ah, so this would remove support for ENV variables like `MONGO_INITDB_ROOT_USERNAME`? I understand why this could be a difficult choice, and I completely understand why you might be worried about how this "represents" mongodb as a whole.
> 
> Setting things up with only a few environment variables is _easy_ after all, and it feels like one of mongo's selling points to new devs is that it's an easy to use database... It's also how other other docker images like `mysql` are ran
> 
> Is there noone on the mongodb who can weigh in on this?

Just to reiterate, I'm personally in favor of removing _all_ the custom stuff like that. I love how easy it is to do but it should be mongo's job to officially support it better. If support for that kind of stuff is what people feel should exist, this should be a thing that uses the core mongo:6 as a baseline with a different thing to suppose customizations. Maybe something like `mongo:6-easy` or something as I've seen other products have different tags to represent the same product with different packaging/rules associated with them.

File: docker-entrypoint.sh [hunk 2]
```diff
@@ -55,362 +53,11 @@ esac
...
-	haveBindIp=
-	if _mongod_hack_have_arg --bind_ip "$@" || _mongod_hack_have_arg --bind_ip_all "$@"; then
-		haveBindIp=1
-	elif _parse_config "$@" && jq --exit-status '.net.bindIp // .net.bindIpAll' "$jsonConfigFile" > /dev/null; then
-		haveBindIp=1
-	fi
-	if [ -z "$haveBindIp" ]; then
-		# so if no "--bind_ip" is specified, let's add "--bind_ip_all"
-		set -- "$@" --bind_ip_all
-	fi
```
Comment (tianon) [inline docker-entrypoint.sh:409]: :facepalm: this part is kind of important :joy:

(rediscovered while testing https://github.com/docker-library/official-images/pull/13006)

**Comment by yosifkit** [comment]
Just to add my thoughts, I don't think we should remove the current entrypoint functionality. While the entrypoint is "hacky" and should probably not be used for production database setup\*, it does make the image much more user friendly for local development purposes.

\* since a production setup is likely much more complex with things like sharding and replicas (like needing an external service to know when and which one should have `rs.init` applied and how to join or remove new replicas). Our script can't currently keep up with all the different production-like configuration setups we have to change or remove in order to start the temporary server for first initialization (like #509).

**Comment by tianon** [comment]
That's fair - feel free to get rid of my commit (that's part of why I left it separate) and come up with an alternative fix :eyes:

**Comment by xianzhuo-sky** [comment]
Local development user want newest image and newest feature, keep up with the mongodb version releases

**Review by tianon — APPROVED** [review]

**Pull request merged**